A Recipe Wild Update

Fusezion opened the pull request:
Oh no, a wild recipe update appeared and forced their way through the entrance. All jokes aside this update has taken far too long to be done and has been restarted far too many times. Lets get a gloss over everything planned for this update and changes made throughout it so far

# Planned Roadmap
- [x] Add a Type for Recipes
- [x] Add an expression for retrieving a recipe via id
- [x] Add additional RecipeUtil methods
- [x] Update Expressions to [Simple]PropertyExpressions
- [x] Enhance existing material choice class expression
- [x] Enhance existing recipe expression classes
- [x] Add recipe group expression 
- [x] Add recipe shape expression (labled no docs, used mainly for debugging)
- [x] Add discovered recipes expression
- [x] Update CondHasDiscoveredRecipe 
- [x] Transfer away from ItemType if applicable, using ItemStack instead 
- [x] Transfer away from String.class to Recipe.class
- [x] Add event-recipe support to select events 
- [x] Move RecipeBookClickEvent to EvtRecipe class 
- [x] Add recipe id expression
- [x] Remove EffRecipeDiscovery *handled via new expression*
- [x] Update EffRemoveRecipe
- [x] Update EffCookingRecipe
- [x] Update EffStonecutterRecipe
- [x] Update EffSmithingRecipe
- [x] Update EffCraftingRecipe

# Breaking Changes
Into the depressing truth of this PR yes there are breaking changes, no there is no real work around. 
**ExprIngredientsOfRecipe.java**
```diff
- [(all [[of] the]|the)] ingredients (for|of) recipe %string%
+ [recipe] ingredients of %recipe%
+ %recipes%'[s] [recipe] ingredients
```
Caused by change to PropertyExpression

**ExprRecipeResult.java**
```diff
- result[s] of recipe[s] [with id[s]] %strings%
+ recipe result[s] of %recipe%
+ %recipes%'[s] recipe result[s]
```
Caused by change to SimplePropertyExpression

**ExprRecipeType.java**
```diff
- recipe type of recipe[s] [with id[s]] %strings%
+ recipe type of %recipe%
+ %recipes%'[s] recipe type
```
Caused by change to SimplePropertyExpression

**ExprMaterialChoice.java**
```diff
- material choice of %itemtypes%
- material choice of [minecraft] tag %minecrafttag%
+ material choice of %itemtypes/minecrafttags%
```
The decision here was to be in favor of allowing multiple choices for material choice
the old system limited to one or the other this one allows combining more making stuff like
```applescript
set {_m} to material choice of minecraft tag of "minecraft:planks", every sword and minecraft block tag of "minecraft:beacon_base_blocks"
```

**CondHasDiscoveredRecipe.java**
```diff
- %players% (has|have) discovered recipe[s] %strings%
- %players% (has|have) not discovered recipe[s] %strings%
+ %players% (has|have) discovered %recipes%
+ %players% (doesn't|does not|do not|don't) have discovered %recipes%
```
Originally planned to make this a PropertyCondition but limitation of arrays and such prevented this, added the classes' syntax 
breaking change caused by update to recipe type

# Additional Notes
While I did agree on attempting to cause zero breaking changes sadly it wasn't likely to happen as such I believe these changes to be better for the future rather than going out of my way to retain current support

Fusezion commented:
If there's any class I forgot to update documentation for please let me know, while I tend to try to recheck before committing I do miss them

Fusezion commented:
Gonna take a break from this for now since I rushed myself too much for remaking this, once I relax I'll go forward and work on the last couple effects then unmark from draft

Fusezion commented:
Last commit accidently included my EffRemoveRecipes class patterns got added nothing was changed, will have other stuff done in the future mistake on my end

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/conditions/CondHasDiscoveredRecipe.java:
```diff
@@ -9,62 +9,55 @@
+                isNegated() ? "has not" : "has",
+                recipes.toString(event,debug));
     }
-
```
This line should not have been removed.
Its generally standard in Java to have a line before the last bracket

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprDiscoveredRecipes.java:
```diff
@@ -0,0 +1,117 @@
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "discovered recipes of " + getExpr().toString(event, debug);
+	}
```
Make sure to add the extra line here too please :)

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeType.java:
```diff
@@ -1,82 +1,39 @@
-
-import java.util.ArrayList;
-import java.util.List;
+import org.eclipse.jdt.annotation.Nullable;
```
I dunno why, but I prefer using jet brains annotations over eclipse.
If you could revert that I'd appreciate it.

I've slowly been trying to use that across all new stuff I add

Fusezion replied:
Never noticed I used an eclipse as I'm on intelij

ShaneBeee replied:
its something that's part of Skript, that's why.
The annotations for the classes/methods were extending from are Eclipse

ShaneBeee reviewed (changes requested):
Just a "few" little changes request.
Mostly formatting things

Fusezion commented:
Currently delayed and no progress, shaped recipes are hard and nearly impossible to really explain

Fusezion commented:
Failing is caused by usage of 2.7.0-beta2 entry stuff

Fusezion commented:
Alrighty so time to cry and post updates about changes since last commit 

1. Recipe Category support has been added as a 1.19 this included a type, expression and option to define in crafting and cooking  effects
2. EffRemoveRecipe has been replaced with EffUnregisterRecipe which includes a few breaking changes in syntax 
firstly this was renamed/replaced in favor of making more sense and we are technically unregistering them
secondly breaking changes
breaking change comes from it's original pattern one where `remove minecraft/custom recipe "bed"` existed it's been removed
3. EffCookingRecipe and EffCraftingRecipe have received mild changes to the pattern formatting making them more readable when on documentation
4. All other register effects have been updated to use systems added to RecipeUtils along with 
5. Removed some unused code in RecipeType
6. RecipeUtil had an update to getRecipeChoice erroring due to skript giving itemtype priority along with added `#getCraftingIngredients` planned more for future addition of `ingredient by key %string% of %recipe%` where I want to add the ability to get ex: `ingredient of key "c" of recipe with id "minecraft:oak_door"` in this case it's `oak plank`
7. recipe group will return null if no group is set instead of sending an empty string 
8. recipe shape is no longer `@NoDoc` and now returns an array with optional `compact` version sending the original design
9. Added shaped recipe section *cause of build failing*, since EffCraftingRecipe gave me a migraine we've abandoned fixing it and instead added a section for registering a shaped recipe to meet your needs 
10. **TODO** update and touch examples to keep up with new and existing syntax changes *mostly only need to for `EffUnregisterRecipe`

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffBrewingRecipe.java:
```diff
@@ -0,0 +1,93 @@
+import org.jetbrains.annotations.Nullable;
+
+@Name("Recipe - Brewing")
+@Description({"Registers a new custom brewing stand recipe",
+        "Input refers to the bottom 3 slots of a brewing stand",
+        "Note: brewing recipes are fairly new within paper, as such support for this is minimal.",
+        "there is currently no way to retrieve all potion recipes or get recipe information like ingredients",
+        "Requires PaperMC 1.18+"})
```
```suggestion
@Description({"Registers a new custom brewing stand recipe.",
        "Input refers to the bottom 3 slots of a brewing stand.",
        "\nNote: brewing recipes are fairly new within Paper, as such support for this is minimal.",
        "There is currently no way to retrieve all potion recipes or get recipe information like ingredients.",
        "Requires PaperMC 1.18+"})
```
Just me being nit picky with formatting.

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffBrewingRecipe.java:
```diff
@@ -0,0 +1,93 @@
+                + " using " + ingredient.toString(event, debug)
+                + " with id " + keyID.toString(event, debug);
+    }
+}
```
```suggestion
}

```
I dunno, but apparently it's common practice to have a blank line at the end. 
Don't ask me why, I just follow the rules 😆

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCookingRecipe.java:
```diff
@@ -41,111 +37,105 @@
         "\tregister new smoking recipe for cooked cod named \"Hot Cod\" using puffer fish with id \"smoking_cod\""})
-@RequiredPlugins("1.13+ for furnaces. 1.14+ for smokers, blast furnaces and campfires.")
-@Since("1.0.0")
+@RequiredPlugins("1.13+ for furnaces. 1.14+ for smokers, blast furnaces and campfires. 1.19+ for Categories")
```
In the description and examples can we please make note/examples of categories.

RequiredPlugins is actually not used by SkriptHub docs, so this will never be seen.
I should have probably removed all usage of this.

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffCraftingRecipe.java:
```diff
@@ -45,42 +43,50 @@
         "\tset {_a} to material choice of every plank",
         "\tregister new shaped recipe for jigsaw block using {_a}, {_a}, {_a}, {_a}, {_a}, {_a}, {_a}, {_a} and {_a} with id \"jigsaw\""})
-@RequiredPlugins("1.13+")
+@RequiredPlugins("Minecraft 1.13+. Minecraft 1.19+ for Categories")
```
Same comment as I stated in cooking recipes. 
Description/examples regarding categories

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffResetRecipes.java:
```diff
@@ -0,0 +1,70 @@
+        return "reset " + getResetType() + " recipes";
+    }
+
+}
```
```suggestion
}

```
silly me again with my extra lines

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffUnregisterRecipe.java:
```diff
@@ -0,0 +1,83 @@
+import org.jetbrains.annotations.Nullable;
+
+@Name("Recipe - Unregister")
+@Description({"Remove a recipe from your server. Recipes can be removed at any time ",
+        "but it is best to do so during a server load event. If a recipe is removed whilst a player is online ",
+        "it will still show up in their recipe book, but they will not be able to craft it. If need be, you can get ",
+        "a list of all recipes by simply typing \"/minecraft:recipe give YourName \" in game.",
+        "You can remove Minecraft recipes, custom recipes and recipes from other plugins. Requires MC 1.13+",
+        "You can remove custom potion recipes by using the potion keyword. Requires PaperMC 1.18+"})
```
```suggestion
@Description({"Remove a recipe from your server. Recipes can be removed at any time",
        "but it is best to do so during a server load event. If a recipe is removed whilst a player is online",
        "it will still show up in their recipe book, but they will not be able to craft it. If need be, you can get",
        "a list of all recipes by simply typing \"/minecraft:recipe give YourName \" in game.",
        "You can remove Minecraft recipes, custom recipes and recipes from other plugins.",
        "You can remove custom potion recipes by using the potion keyword. Potions requires PaperMC 1.18+"})
```
Just some formatting stuff.
SkriptHub docs takes care of spaces.

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/effects/EffUnregisterRecipe.java:
```diff
@@ -0,0 +1,83 @@
+        return "unregister" + (removePotion ? " potion " : "") + "recipe(s)" + recipes.toString(event, debug);
+    }
+
+}
```
```suggestion
}

```
I think you know what im going to say ;)

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeCategory.java:
```diff
@@ -0,0 +1,52 @@
+import org.eclipse.jdt.annotation.Nullable;
+
+@Name("Recipe - Category")
+@Description("Returns the category of a valid recipe")
```
```suggestion
@Description("Returns the category of a valid recipe. Requires Minecraft 1.19+")
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeCategory.java:
```diff
@@ -0,0 +1,52 @@
+        return "recipe category";
+    }
+
+}
```
```suggestion
}

```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeId.java:
```diff
@@ -0,0 +1,44 @@
+import org.jetbrains.annotations.Nullable;
+
+@Name("Recipe - Recipe ID")
+@Description({"Gets the registered identifier of recipes",
+        "Requires Minecraft 1.13+"})
```
```suggestion
@Description("Gets the registered identifier of recipes.")
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeId.java:
```diff
@@ -0,0 +1,44 @@
+@Name("Recipe - Recipe ID")
+@Description({"Gets the registered identifier of recipes",
+        "Requires Minecraft 1.13+"})
+@Examples({"set {_recipe} to random element out of recipes",
+        "send recipe id of {_recipe}"
+})
```
```suggestion
@Examples({"set {_recipe} to random element out of recipes",
        "send recipe id of {_recipe}"})
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeId.java:
```diff
@@ -0,0 +1,44 @@
+        return "recipe id";
+    }
+
+}
```
```suggestion
}

```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeShape.java:
```diff
@@ -0,0 +1,71 @@
+
+// TODO: add recipe ingredient by shape key, makes support for custom shape more usable in long term
+@Name("Recipes - Recipe Shape")
+@Description({"Gets the registered shape of a Shaped Recipe",
+        "Recipe shape will return an array of the shape, while compacted will return a string in an easier to read format.",
+        "Requires Minecraft 1.13+"})
```
```suggestion
@Description({"Gets the registered shape of a Shaped Recipe.",
        "Recipe shape will return an array of the shape, while compacted will return a string in an easier to read format."})
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeShape.java:
```diff
@@ -0,0 +1,71 @@
+        return (this.isCompact ? "compacted " : "") + "recipe shape of " + getExpr().toString(event, debug);
+    }
+
+}
```
```suggestion
}

```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeWithId.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.List;
+
+@Name("Recipe - Recipe with Id")
+@Description({"Gets a minecraft recipe based off a given recipe id. Requires Minecraft 1.13+",
```
```suggestion
@Description({"Gets a minecraft recipe based off a given recipe id.",
```
I got all the way down here and then realized the "requires mc 1.13" isn't needed, since this version won't work on Skript 1.13 anyways. 
Now I have to scroll up and change some stuff haha

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeWithId.java:
```diff
@@ -0,0 +1,70 @@
+        return "recipe from id " + recipeIds.toString(event, debug);
+    }
+
+}
```
```suggestion
}

```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/sections/SecShapedRecipe.java:
```diff
@@ -0,0 +1,183 @@
+import java.util.List;
+
+@Name("Recipes - Advanced Shaped Recipe")
+@Description({"Ceates a new shaped recipe using sections and entries",
+        "When using the shape entry, it is required to follow a rectangular shape in example 'ab, cde' will not work while 'abc, def' will work"})
```
```suggestion
@Description({"Ceates a new shaped recipe using sections and entries.",
        "When using the shape entry, it is required to follow a rectangular shape in example 'ab, cde' will not work while 'abc, def' will work".})
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/sections/SecShapedRecipe.java:
```diff
@@ -0,0 +1,183 @@
+        return "register shaped recipe for " + result.toString(event, debug) + " with id " + keyID.toString(event, debug);
+    }
+
+}
```
```suggestion
}

```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/type/Types.java:
```diff
@@ -50,6 +54,35 @@ public String getVariableNamePattern() {
+        Classes.registerClass(new ClassInfo<>(Recipe.class, "recipe")
+                .name("Recipes - Recipe")
+                .user("recipes?")
+                .description("Represents a minecraft recipe which is used to collect information",
+                        "Requires Minecraft 1.13+")
```
```suggestion
                .description("Represents a Minecraft recipe which is used to collect information.")
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/type/Types.java:
```diff
@@ -58,6 +91,27 @@ public String getVariableNamePattern() {
+            Classes.registerClass(new ClassInfo<>(CookingBookCategory.class, "cookingcategory")
+                    .user("cooking ?categor(y|ies)")
+                    .name("Recipes - Cooking Category")
+                    .description("Represents the type of cooking recipe book categories")
```
```suggestion
                    .description("Represents the type of cooking recipe book categories.")
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/type/Types.java:
```diff
@@ -58,6 +91,27 @@ public String getVariableNamePattern() {
+            Classes.registerClass(new ClassInfo<>(CraftingBookCategory.class, "craftingcategory")
+                    .user("crafting ?categor(y|ies)")
+                    .name("Recipes - Crafting Category")
+                    .description("Represents the type of crafting recipe book categories")
```
```suggestion
                    .description("Represents the type of crafting recipe book categories.")
```

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprDiscoveredRecipes.java:
```diff
@@ -0,0 +1,114 @@
+import java.util.List;
+
+@Name("Recipes - Discovered")
+@Description({"Get and modify the discovered recipes of a player",
+        "Requires minecraft 1.13+"})
```
```suggestion
@Description("Get and modify the discovered recipes of a player.")
```
I made not of this down further, but since this is a new class, SkBee doesn't support 1.13 anyways.

ShaneBeee commented on src/main/java/com/shanebeestudios/skbee/elements/recipe/expressions/ExprRecipeGroup.java:
```diff
@@ -0,0 +1,54 @@
+import org.jetbrains.annotations.Nullable;
+
+@Name("Recipe - Group of Recipe")
+@Description({"Get the group of a recipe, not all recipes support groups.",
+        "Requires Minecraft 1.13+"})
```
```suggestion
@Description("Get the group of a recipe, not all recipes support groups.")
```

ShaneBeee reviewed (changes requested):
Ok just a FEW little tiny changes

Fusezion commented:
Closing in favor of slowly releasing the recipe update vs. making one large pr of everything making it hard to review. 
New plan can be found at #418 which will be updated with progress as we go on
